internal/provider/provider_framework.go

```diff
@@ -0,0 +1,595 @@
+						},
+						"key_secret": schema.StringAttribute{
+							Optional: true,
+							Validators: []validator.String{
+								stringvalidator.ConflictsWith(path.MatchRelative().AtParent().AtName("gssapi")),
+								stringvalidator.AlsoRequires(
+									path.MatchRelative().AtParent().AtName("key_name"),
+									path.MatchRelative().AtParent().AtName("key_algorithm"),
+								),
+							},
+							Description: "Required if `key_name` is set\nA Base64-encoded string containing the " +
+								"shared secret to be used for TSIG.",
+						},
+					},
+					Blocks: map[string]schema.Block{
+						"gssapi": schema.ListNestedBlock{
+							Validators: []validator.List{
+								listvalidator.SizeAtMost(1),
+								listvalidator.ConflictsWith(
+									path.MatchRelative().AtParent().AtName("key_name"),
+									path.MatchRelative().AtParent().AtName("key_algorithm"),
+									path.MatchRelative().AtParent().AtName("key_algorithm"),
+								),
+							},
+							Description: "A `gssapi` block. Only one `gssapi` block may be in the configuration. " +
+								"Conflicts with use of `key_name`, `key_algorithm` and `key_secret`.",
+							NestedObject: schema.NestedBlockObject{
+								Attributes: map[string]schema.Attribute{
+									"realm": schema.StringAttribute{
+										Required:    true,
+										Description: "The Kerberos realm or Active Directory domain.",
+									},
+									"username": schema.StringAttribute{
+										Optional: true,
+										Description: "The name of the user to authenticate as. If not set the current " +
+											"user session will be used.",
+									},
+									"password": schema.StringAttribute{
+										Optional: true,
+										Validators: []validator.String{
+											stringvalidator.ConflictsWith(path.MatchRelative().AtParent().AtName("keytab")),
+											stringvalidator.AlsoRequires(path.MatchRelative().AtParent().AtName("username")),
+										},
+										Sensitive: true,
+										Description: "This or `keytab` is required if `username` is set. The matching " +
+											"password for `username`.",
+									},
+									"keytab": schema.StringAttribute{
+										Optional: true,
+										Validators: []validator.String{
+											stringvalidator.ConflictsWith(path.MatchRelative().AtParent().AtName("password")),
+											stringvalidator.AlsoRequires(path.MatchRelative().AtParent().AtName("username")),
+										},
+										Description: "This or `password` is required if `username` is set, not " +
+											"supported on Windows. The path to a keytab file containing a key for " +
+											"`username`.",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (p *dnsProvider) Configure(ctx context.Context, req provider.ConfigureRequest, resp *provider.ConfigureResponse) {
+	var providerConfig providerModel
+	var providerUpdateConfig providerUpdateModel
+	var providerGssapiConfig providerGssapiModel
```

Comment:
I sent a note outside of GitHub with more details but these should include `[]` since they are lists. It didn't appear there was acceptance testing covering "successful" Terraform configurations like:

```terraform
provider "dns" {
  update {
    server = "127.0.0.1"
  }
}
```